File: elixir/lib/symphony/prompt_builder.ex
defmodule Symphony.PromptBuilder do
  @moduledoc """
  Renders Liquid (Solid) templates with issue fields as context.
  """

  @doc "Render a Liquid template string with issue context."
  def render(template, issue) do
    context = build_context(issue)

    with {:ok, parsed} <- Solid.parse(template),
         {:ok, iodata} <- Solid.render(parsed, context) do
      {:ok, IO.iodata_to_binary(iodata)}
    else
      {:error, reason} -> {:error, {:template_error, reason}}
      {:error, errors, _} -> {:error, {:template_error, errors}}
    end
  end

  defp build_context(issue) do
    %{
      "identifier" => Symphony.GitHub.Issue.identifier(issue),
      "number" => issue.number,
      "title" => issue.title || "",
      "description" => issue.body || "",
      "body" => issue.body || "",
      "state" => issue.state || "",
      "labels" => issue.labels || [],
      "assignees" => issue.assignees || [],
      "url" => issue.url || "",
      "priority" => to_string(issue.priority),
      "blockers" => Enum.map(issue.blockers, &to_string/1),
      "repo" => Symphony.Config.tracker_repo() || ""
    }
  end
end

File: src/prompt-builder.ts
import { Effect, Data } from "effect"
import { Liquid } from "liquidjs"
import { Config } from "./config.js"
import { type Issue, identifier } from "./github/issue.js"

export class TemplateError extends Data.TaggedError("TemplateError")<{
  readonly reason: string
}> {}

const engine = new Liquid()

/** Render a Liquid template with issue context. */
export function render(
  template: string,
  issue: Issue,
  attempt: number | null = null,
): Effect.Effect<string, TemplateError, Config> {
  return Effect.flatMap(Config, (config) => {
    const context: Record<string, unknown> = {
      issue: {
        identifier: identifier(issue),
        number: issue.number,
        title: issue.title,
        description: issue.body,
        body: issue.body,
        state: issue.state,
        labels: issue.labels,
        assignees: issue.assignees,
        url: issue.url,
        priority: issue.priority,
        blockers: issue.blockers.map(String),
      },
      attempt,
      repo: config.trackerRepo,
      // Keep top-level aliases for backward compatibility with existing templates
      identifier: identifier(issue),
      number: issue.number,
      title: issue.title,
      description: issue.body,
      body: issue.body,
      state: issue.state,
      labels: issue.labels,
      assignees: issue.assignees,
      url: issue.url,
      priority: issue.priority,
      blockers: issue.blockers.map(String),
    }

    return Effect.tryPromise({
      try: () => engine.parseAndRender(template, context),
      catch: (err) =>
        new TemplateError({
          reason: err instanceof Error ? err.message : String(err),
        }),
    })
  })
}
